### ISSUE TITLE
Add support for controlling multiple bindings with the same name.

### ISSUE THREAD

**Issue by mtj8033** [spring-cloud/spring-cloud-stream#1568]
```

	private Binding<?> locateBinding(String name) {
		Stream<Binding<?>> bindings = Stream.concat(this.gatherInputBindings().stream(), this.gatherOutputBindings().stream());
		return bindings
			.filter(binding -> name.equals(binding.getName()))
			.findFirst()
			.orElse(null);
	}

```

Currently this code finds the first one given a particular name, however if I have a Kafka producer and a Kafka Streams consumer in the same codebase it ends up having two with the same name so there isn't a way to reach whichever one is second to pause/resume it/check status with the get endpoint/etc. The main /actuator/bindings endpoint correctly shows the status of all of them though

**Comment by olegz** [comment]
I am not ready to classify this as a big since as you pointed out in the code this is intentional. That said, it is lacking functionality to be able to handle the scenario you describe

**Comment by olegz** [comment]
So, this is a bit tricky. 
1.  I am questioning the value (outside of testing) of the Producer and Consumer in the same application. I mean if you producing something that has to be consumed/processed in the same app, why not just call the processing method directly? Why use remote messaging broker like Kafka/Rabbit etc.?
2.  If there is  proper justification, then we need some unique identifier for the Binding since we can not rely on the name. If so, should such id be as simple as newly generated id at the star time or should it be the type of id that is always the same and so on. . .

Basically this is not even an enhancement, rather a new feature which largely depends on answer to question 1.

**Comment by mtj8033** [comment]
For 1. Testing is probably the main use case and the only real one I have currently, but I could see someone using it for throttling purposes, for example one message spawning off into 500 individual messages that are only consumed 5 at a time or something

One thing to note is that it does feel a little misleading for the get call to return all the binders, but the individual calls to only be allowed to impact one of them. Thoughts on logging a warning if there is more than one with the same name?

**Comment by olegz** [comment]
I keep struggling with the concept of supporting something that is only used in testing. For throttling case there are other ways of accomplishing it.
This means that outside of testing this is not a realistic scenario as you essentially go against the framework design. If you want to wire several components together via channels in a single JVM for testing purposes then create a test SI configuration. . . Spring Cloud Stream delegates to SI anyway, so you're not even changing the invocation model (send/receive to/from channel), just the activation (rabbit/kafka).

I do not see any tangible actions other then further discussion on this issue.

**Comment by mtj8033** [comment]
Do all the binders use the destination value as the name of their binding? So if I have a destination "workQueue" for rabbit and kafka will they be named the same?

**Comment by olegz** [comment]
@mtj8033 yes, but the bigger point why/when (outside of testing) would this be useful - this being binding Consumer and Producer to the same destination in a single microservice?

**Comment by mtj8033** [comment]
Well the reason I was asking the last question is because then its not even the same actual destination, just the same name.
In this example:
https://docs.spring.io/spring-cloud-stream/docs/current/reference/htmlsingle/#multiple-systems

If the destination name was the same wouldn't they hit the same issue? So I don't know maybe you wanted to read all messages from workQueue in rabbit and write them out into a topic named workQueue in kafka, you'd run into a similar issue if you tried to pause no? Or replicate messages from one rabbit instance to another to facilitate a transition period

**Comment by olegz** [comment]
Good point on multi-binder. . . need to think

**Comment by kdavisk6** [comment]
In my use case, we are using the same Kinesis Stream for inbound and outbound channels.  This is being done because we are sharing a larger Event stream with other microservices.  When using the Kinesis Binder, this works, we can use the same destination name for our INPUT and OUTPUT stream respectively.  In our case, it is the same system and it works.  Could this be something specific to the Kafka binder?

**Comment by olegz** [comment]
I am still struggling to understand. 
If you have one service that produces the data and another that consumes the data and they live within the same JVM, why do they need a remote system such as Kafka, Kinesis etc to communicate with one another?

**Comment by kdavisk6** [comment]
My apologies, in our case, they are two different JVM's.  Sorry the confusion.

**Comment by olegz** [comment]
These seem to be related #1476

**Comment by olegz** [comment]
Resolved with https://github.com/olegz/spring-cloud-stream/commit/144627c4355b51b81ace7c4254e3a8b68ba21a3d
